tests/unit_tests/test_matchmaker_queue_algorithm.py

```diff
@@ -7,32 +9,112 @@
 def p(player_factory):
     def make(mean: int, deviation: int, ladder_games: int=config.NEWBIE_MIN_GAMES+1, name=None):
         """Make a player with the given ratings"""
-        return player_factory(ladder_rating=(mean, deviation),
-                              ladder_games=ladder_games)
+        player = player_factory(
+            ladder_rating=(mean, deviation),
+            ladder_games=ladder_games,
+            with_lobby_connection=False,
+        )
+        return player
     return make
 
 
-def test_build_sparse_matching_graph(p):
+def add_graph_edge_weights(graph) -> algorithm.WeightedGraph:
+    return {
+        s1: [(s2, s1.quality_with(s2)) for s2 in edges]
+        for s1, edges in graph.items()
+    }
+
+
+def test_build_full_matching_graph(p):
     s1 = Search([p(1500, 64, ladder_games=20)])
     s2 = Search([p(1500, 63, ladder_games=20)])
     s3 = Search([p(1600, 75, ladder_games=50)])
     searches = [s1, s2, s3]
 
-    ranks = algorithm._MatchingGraph.build_sparse(searches)
+    ranks = algorithm._MatchingGraph.build_full(searches)
 
-    assert ranks == {
+    assert ranks == add_graph_edge_weights({
+        s1: [s3, s2],
+        s2: [s3, s1],
+        s3: [s1, s2]
+    })
+
+
+def test_build_matching_graph_different_ranks(p):
+    s1 = Search([p(1500, 64, ladder_games=20)])
+    s2 = Search([p(200, 63, ladder_games=20)])
+    searches = [s1, s2]
+
+    ranks1 = algorithm._MatchingGraph.build_full(searches)
+    ranks2 = algorithm._MatchingGraph.build_fast(searches)
+
+    empty_graph = add_graph_edge_weights({
+        s1: [],
+        s2: [],
+    })
+
+    assert ranks1 == empty_graph
+    assert ranks2 == empty_graph
+
+
+def test_build_fast_matching_graph(p):
+    s1 = Search([p(1500, 64, ladder_games=20)])
+    s2 = Search([p(1500, 63, ladder_games=20)])
+    s3 = Search([p(1600, 75, ladder_games=50)])
+    searches = [s1, s2, s3]
+
+    ranks = algorithm._MatchingGraph.build_fast(searches)
+
+    import pprint
+
+    pp = pprint.PrettyPrinter(indent=1)
+    pp.pprint(ranks)
+    assert ranks == add_graph_edge_weights({
         s1: [s3, s2],
         s2: [s3, s1],
         s3: [s1, s2]
+    })
+
+
+def test_remove_isolated(p):
+    s1 = Search([p(1500, 64, ladder_games=20)])
+    s2 = Search([p(1500, 63, ladder_games=20)])
+    s3 = Search([p(1600, 75, ladder_games=50)])
+    ranks = add_graph_edge_weights({
```

Comment:
This example is a bit weird because it is not undirected (or not symmetric, if you want).
i.e. there is an edge s1 -> s2 but not s2 -> s1.
In case we ever expect to have directed graphs I would expect `remove_isolated` to only remove nodes which have neither in- nor outgoing edges, so I would rather document that expectation.
That is,
```python
  s1: [s3],
  s2: [],
  s3: [s1],
```

Reply:
Mmmm. Good catch. Maybe I'll rename the function or add to the docstring that it assumes edges are bidirectional. Because such a graph should never be created so there's no point in adding extra logic to handle that case if we don't need it.

Reply:
Decided to just add a description to the docstring since I don't really expect our matchmaking to ever need directed graphs. I was actually thinking this function might not be necessary at all. I doubt that it makes a performance impact to remove isolated vertices before running the matching algorithm since they will just be skipped anyways.